Dev/vk

wasertech opened the pull request:
### Proposed changes / Предлагаемые изменения

This pull request introduces significant improvements to cross-platform build support, especially for Linux and Vulkan, and enhances package management and renderer flexibility. The most important changes include improved compiler selection in CMake, expanded Linux and Vulkan support, and more robust CI workflows.

**Build system and cross-platform improvements:**

* Added logic in `CMakeLists.txt` to auto-select compilers based on environment variables or system availability, improving build reliability on different systems.
* Added new CMake options for Vulkan renderer support (`IXRAY_USE_VK`, `IXRAY_VK_STRICT_NO_D3D`) and enabled more flexible renderer configuration.

**CI/CD and workflow enhancements:**

* Updated `.github/workflows/build-engine.yml` to add Linux build and artifact upload steps, ensuring Linux binaries and libraries are built and archived in CI. [[1]](diffhunk://#diff-3831ad170138a002239ab20151a12a239e238d0488fabe2f35886519e5000e38R273-R280) [[2]](diffhunk://#diff-3831ad170138a002239ab20151a12a239e238d0488fabe2f35886519e5000e38R298-R307) [[3]](diffhunk://#diff-3831ad170138a002239ab20151a12a239e238d0488fabe2f35886519e5000e38R326-R343)
* Added Linux-specific build presets in `CMakePresets.json` for streamlined local and CI builds.

**Third-party package management:**

* Improved `cmake/linux/packages.cmake` to fetch or locate required dependencies (yaml-cpp, TBB, LZO, nvtt, lunasvg), and added conditional fetching of Vulkan SDK components when Vulkan renderer is enabled. [[1]](diffhunk://#diff-5560eb9dbfb7c9f5bbef46c8f3431e980bbd0c78e869cc4e242f49604435c46cL3-R37) [[2]](diffhunk://#diff-5560eb9dbfb7c9f5bbef46c8f3431e980bbd0c78e869cc4e242f49604435c46cL41-R129)

**Renderer and platform abstraction:**

* Refactored `src/Layers/xrRender/FVF.h` to allow compiling without Direct3D dependencies, using fallback types and dummy constants for non-D3D platforms or strict Vulkan builds.
* Updated device creation/destruction logic in `src/xrEngine/Device_create_render.cpp` to support Vulkan alongside DX9/DX11, with new Vulkan-specific initialization and cleanup functions. [[1]](diffhunk://#diff-1233ef6c6ab5e04ca12ec719060548f1914d8228966e554b8b1c0e6053acad9bR173-R177) [[2]](diffhunk://#diff-1233ef6c6ab5e04ca12ec719060548f1914d8228966e554b8b1c0e6053acad9bR348-R353) [[3]](diffhunk://#diff-1233ef6c6ab5e04ca12ec719060548f1914d8228966e554b8b1c0e6053acad9bR375-R387)

**Miscellaneous:**

* Minor bugfix in `src/Layers/xrRender/R_DStreams.h` to remove duplicate member declarations.

These changes collectively make the engine more portable, easier to build across different environments, and ready for Vulkan-based rendering.

### Associated issues / Ассоциированные проблемы

https://github.com/ixray-team/ixray-1.6-stcop/pull/185#issuecomment-3263620153

### Testing / Тестирование

<!--
Please write 'x' letter in '[]' to agree. For example:
Пожалуйста, напишите букву 'x' в '[]' для подтверждения. Например:

- [x] Manual testing / Ручное тестирование
-->

- [ ] Manual testing / Ручное тестирование

Don't have ubuntu... building on Arch based is another can of worms.

### Agreements / Согласия

- I agree to follow this project's [__Contributing Guidelines__](../CONTRIBUTING.md) /
Я согласен(-на) следовать [__Рекомендациям по внесению вклада в этот проект__](../CONTRIBUTING.rus.md)
- I agree to follow this project's [__Code of Conduct__](../CODE_OF_CONDUCT.md) /
Я согласен(-на) следовать [__Кодексу поведения этого проекта__](../CODE_OF_CONDUCT.rus.md)

Copilot committed:
Implement basic Vulkan renderer infrastructure

Co-authored-by: wasertech <79070834+wasertech@users.noreply.github.com>

Copilot committed:
Complete Vulkan renderer implementation with verification and documentation

Co-authored-by: wasertech <79070834+wasertech@users.noreply.github.com>

Copilot committed:
Implement traditional Vulkan render passes instead of dynamic rendering

Co-authored-by: wasertech <79070834+wasertech@users.noreply.github.com>

wasertech committed:
typo fix :)

wasertech committed:
Added Vulkan SDK installation to build workflows and updated Vulkan rendering integration check with shader path.

wasertech committed:
identified and fixed the CMake error with Vulkan in CI

wasertech committed:
Added Vulkan SDK installation to build workflows for editors, server, and utilities, with error handling and a fallback method via Chocolatey.

wasertech committed:
Replaces manual Vulkan SDK installation with a GitHub action for easier and faster setup in build workflows.

wasertech committed:
Added Vulkan SDK version management to build workflows and implemented fallbacks for Vulkan loader discovery in CMake.

wasertech committed:
Added Vulkan SDK installation and configuration to build workflows for editors, server, plugins, and utilities, with cache management and error checking.

wasertech committed:
Refactor ResizeBuffers function for cleaner case handling of API levels

wasertech committed:
Updated Vulkan validation packages for Ubuntu 24.04 by replacing *-dev packages with separate packages.

wasertech committed:
Added fallback implementation for vertex and index streams in Linux builds (Vulkan/OpenGL) and refactored _VertexStream and _IndexStream classes for improved memory management.

wasertech committed:
Added fallback values ​​for Direct3D FVF constants in Linux builds (Vulkan/OpenGL) to ensure shared code compiles.

wasertech committed:
Added fallback paths for engine base types in FVF.h and improved memory management in R_DStreams.h for Linux builds (Vulkan/OpenGL).

wasertech committed:
Added Vulkan SDK caching and installation for Linux builds, with export of appropriate environment variables.

wasertech committed:
Added support for installing and exporting the Vulkan SDK environment for Windows in the build workflow.

wasertech committed:
Added support for Vulkan system packages as a fallback for Linux builds and enabled the use of Vulkan in the engine configuration.

wasertech committed:
Added Vulkan support to CMake configuration, including options to enable Vulkan rendering and improved error messages for Vulkan SDK configuration.

wasertech committed:
Added support for Vulkan on Linux by installing the necessary system packages and removing the steps to cache and install the Vulkan SDK.

wasertech committed:
Updated Vulkan SDK version to 1.3.290.0 for Windows builds and adjusted CMake cache configuration.

wasertech committed:
Added IXRAY_VK_STRICT_NO_D3D option to force the Vulkan compiler to not depend on Direct3D, even on Windows. Updated build definitions to handle this option in CMake files and headers.

wasertech committed:
Added new CMake presets for configuring the Engine on Linux, including options for Debug, RelWithDebInfo, and Release modes.

wasertech committed:
Disabled IXRAY_USE_R1 and IXRAY_USE_R2 options, enabled IXRAY_USE_VK in CMake configuration.

wasertech committed:
Added support for compiling on Linux: Install build-essential, configure C/C++ compilers via environment variables, and validate the LuaJIT library with a symlink if necessary. Updated the build error reference file.

wasertech committed:
Added Mono and NuGet installation for Linux builds, updated the NuGet installation command, and improved error handling when restoring packages.

wasertech committed:
Added export of the VULKAN_SDK environment variable for Linux builds, defining the path to the system Vulkan headers and libraries.

wasertech committed:
Added Vulkan configuration check and updated CMake options to disable R1 and R2 while enabling IXRAY_FORCE_NO_D3D. Installed libvulkan-dev for Linux builds.

wasertech committed:
Added auto-security logic to check for availability of Direct3D headers in FVF.h and R_DStreams.h files, ensuring proper fallback if needed.

wasertech committed:
Added Direct3D fallback logic for non-D3D setups, including minimal definitions and adjustments in header files to ensure compatibility with Vulkan.

wasertech committed:
Added Mono and NuGet installation for Linux targets, with a wrapper script and NuGet package restore logic.

wasertech committed:
Updated dependencies for Linux build: added libtbb-dev and liblzo2-dev, and changed TBB fetching logic via FetchContent.

wasertech committed:
Added recovery logic for several Linux dependencies, including Vulkan and its components, with improved status messages for each step.

wh1t3lord commented:
@wasertech did you try to install WSL Ubuntu and test your PR on that platform? Because WSL works under Windows

wasertech commented:
> @wasertech did you try to install WSL Ubuntu and test your PR on that platform? Because WSL works under Windows

```zsh
❯ fastfetch
██████████████████  ████████    waser@Destinee
██████████████████  ████████    --------------
██████████████████  ████████    OS: Manjaro Linux x86_64
██████████████████  ████████    Kernel: Linux 6.15.11-2-MANJARO
████████            ████████    Uptime: 1 hour, 41 mins
████████  ████████  ████████    Packages: 2122 (pacman)[stable]
████████  ████████  ████████    Shell: zsh 5.9
████████  ████████  ████████    Display (Mi Monitor): 3440x1440 @ 144 Hz in 34" [External]
████████  ████████  ████████    DE: GNOME 48.4
████████  ████████  ████████    WM: Mutter (Wayland)
████████  ████████  ████████    WM Theme: adw-gtk3-dark
████████  ████████  ████████    Theme: adw-gtk3-dark [GTK2/3/4]
████████  ████████  ████████    Icons: Papirus-Dark-Maia [GTK2/3/4]
████████  ████████  ████████    Font: Noto Sans (11pt) [GTK2/3/4]
                                Cursor: Bibata-Original-Classic (24px)
                                Terminal: tmux next-3.4
                                CPU: AMD Ryzen Threadripper 2920X (24) @ 3.50 GHz
                                GPU 1: NVIDIA TITAN RTX
                                GPU 2: NVIDIA TITAN RTX
                                Memory: 13.38 GiB / 94.26 GiB (14%)
                                Swap: Disabled
                                Disk (/): 262.80 GiB / 392.12 GiB (67%) - ext4
                                Disk (/home/waser): 672.54 GiB / 915.82 GiB (73%) - ext4
                                Disk (/mnt/Archives): 1.32 TiB / 5.41 TiB (24%) - ext4
                                Disk (/mnt/Data_II): 3.11 TiB / 3.58 TiB (87%) - ext4
                                Disk (/mnt/Jeux): 374.39 GiB / 420.68 GiB (89%) - ext4
                                Locale: fr_CH.UTF-8

                                                        
                                                        

```

But I am not using Microsoft's OS... I use an Arch based distro. I have bleeding edge deps. The engine needs serious work before we are able to build it on Arch... ubuntu-latest is a nice intermediate target for now.

wh1t3lord commented:
@wasertech my bad I didn't know that you have native linux distro 😄

wasertech committed:
rm duplicates

wasertech commented:
On arch it crashes like so if you are curious:
```zsh
rm -rf build/x64/Engine-Linux                                                                                                                                                                                                                                                           ─╯
cmake --preset Engine-x64-Linux -DIXRAY_USE_VK=ON -DIXRAY_USE_R1=OFF -DIXRAY_USE_R2=OFF -DIXRAY_FORCE_NO_D3D=ON
cmake --build --preset Engine-x64-Linux-Debug --target luabind
...
-- [IXRAY] XR_FORCE_NO_D3D is active (Direct3D paths disabled).
-- Vulkan include dirs: /usr/include
-- Vulkan libraries: /lib/libvulkan.so
-- Configuring incomplete, errors occurred!
ninja: error: loading 'build-Debug.ninja': No such file or directory
```

I am sure I am missing a gazillion things but I am tired and hungry so I'll got touch some grass for a while.

Feel free to merge this PR already, I'll sync my branch later.

wasertech committed:
Improved nvtt dependency handling for Linux: added nvtt_SOURCE_DIR definition check and defensive inclusion of FetchContent macros.

wasertech committed:
Added Vulkan and CUDA dependency management for Windows and Linux, with configuration options for TBB and NVTT. Improved status messages for dependency retrieval.

wasertech committed:
Improved NVTT handling for Linux: added an option to disable NVTT, checked for the presence of the nvtt target, and managed include paths.

wasertech committed:
Added CUDA toolkit installation for Windows and Linux to the build workflow. Changed OptiX search so it is not required.

wasertech committed:
Added environment variable management for Vulkan and NVTT in the build workflow. Improved CMake cache for builds.

wasertech committed:
mv script to util/

wasertech committed:
mv docs to doc/

wasertech committed:
Add initial PKGBUILD, .SRCINFO, and build scripts for IX-Ray 1.6 STCOP on Arch

- Created a comprehensive PKGBUILD file with dependencies, build configurations, and installation instructions.
- Added .SRCINFO for package metadata.
- Implemented build and test scripts to facilitate building and validating the package on Arch Linux.
- Updated .gitignore to exclude build artifacts and temporary files.
- Included detailed README-ARCH.md for build instructions specific to Arch Linux.

wasertech commented:
Started looking into local arch builds. I get stuck same place as windows so it's a good thing since it's (mostly) functional. I can now start to build and test locally. which is always nice.

```
❯ grep -RniI -e 'error:' -e 'cmake error' -e 'ninja:' build/x64/Engine-Linux 2>/dev/null | head -n 40
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:4318:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-kZH6LP/CheckSymbolExists.c:26:19: error: use of undeclared identifier '_i64toa'
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:4322:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:4380:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-bdOzRr/CheckSymbolExists.c:26:19: error: use of undeclared identifier 'itoa'
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:4384:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:4582:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-tDlhyc/CheckSymbolExists.c:26:19: error: use of undeclared identifier '_ltoa'
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:4586:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:5120:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-OM6nGR/CheckSymbolExists.c:26:19: error: use of undeclared identifier 'sqr'; did you mean 'sqrt'?
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:5128:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:5522:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-DZtTwd/CheckSymbolExists.c:26:19: error: use of undeclared identifier 'strnstr'; did you mean 'strstr'?
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:5530:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:6652:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-jWhEJF/CheckSymbolExists.c:3:10: fatal error: 'sys/sysctl.h' file not found
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:6656:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:6714:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-N0F8xe/CheckSymbolExists.c:8:19: error: use of undeclared identifier 'elf_aux_info'
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:6718:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:6862:        clang: error: linker command failed with exit code 1 (use -v to see invocation)
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:6863:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:25557:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-Y7MOnG/src.c:2:18: fatal error: 'GLES/gl.h' file not found
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:25561:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:27581:        /home/waser/Projets/ixray-1.6-stcop/build/x64/Engine-Linux/CMakeFiles/CMakeScratch/TryCompile-4oHLFZ/CheckIncludeFile.c:1:10: fatal error: 'pthread_np.h' file not found
build/x64/Engine-Linux/CMakeFiles/CMakeConfigureLog.yaml:27585:        ninja: build stopped: subcommand failed.
build/x64/Engine-Linux/dep/amd_ags_sdk/inc/amd_ags.h:883:/// D3D11 ERROR: ID3D11Device::RemoveDevice: Device removal has been triggered for the following reason (DXGI_ERROR_DEVICE_HUNG: The Device took an unreasonable amount of time to execute its commands, or the hardware crashed/hung. As a result, the TDR (Timeout Detection and Recovery) mechanism has been triggered. The current Device Context was executing commands when the hang occurred. The application may want to respawn and fallback to less aggressive use of the display hardware). [ EXECUTION ERROR #378: DEVICE_REMOVAL_PROCESS_AT_FAULT]
build/x64/Engine-Linux/dep/discord_gamesdk/examples/csharp/Program.cs:190:                Console.WriteLine("user fetch error: {0}", result);
build/x64/Engine-Linux/CMakeCache.txt:825:SDL_WERROR:BOOL=OFF
```

Copilot committed:
Initial plan

Copilot committed:
Fix luabind compilation issues for Linux build

Co-authored-by: wasertech <79070834+wasertech@users.noreply.github.com>

Copilot committed:
Fix additional luabind compilation issues for Linux build

Co-authored-by: wasertech <79070834+wasertech@users.noreply.github.com>

Copilot committed:
Major progress: Fix platform dependencies and core library compilation

Co-authored-by: wasertech <79070834+wasertech@users.noreply.github.com>

Copilot committed:
Implement missing platform-specific functions for Linux

Co-authored-by: wasertech <79070834+wasertech@users.noreply.github.com>

Copilot committed:
Fix Windows build and cross-platform CMake configuration issues

- Fix incorrect include path for packages-vulkan.cmake in Windows and Linux builds
- Use CMAKE_CURRENT_LIST_DIR instead of CMAKE_CURRENT_SOURCE_DIR for proper relative paths
- Add case-sensitivity symlinks (Linux -> linux, Windows -> windows) for cross-platform compatibility
- Resolve CMake configuration errors that were preventing Windows CI builds

Co-authored-by: wasertech <79070834+wasertech@users.noreply.github.com>

wasertech committed:
Merge pull request #4 from wasertech/copilot/fix-3

Enable Vulkan renderer compilation on Linux and Windows by fixing platform-specific dependencies and build configuration issues

wasertech committed:
rm dup

wasertech committed:
Improved NuGet package restoration and updated paths in build scripts

wasertech committed:
Enhance CMake configurations and add syntax checking script

- Updated clang.cmake to conditionally include libc++ based on header availability.
- Refactored CMakeLists.txt files across various modules to improve platform-specific library linking for DirectX and audio libraries.
- Introduced a new check-syntax.sh script to validate C++ syntax across the project using clang++, logging any errors found.
- Improved build.sh to log build output to a file for better traceability.

wasertech committed:
Improved syntax checking script to include .h files and run checks in parallel

wasertech committed:
Added the --keep-logs option to preserve existing error logs when checking C++ syntax and header files. Improved error handling and informational messages.

wasertech committed:
Refactor syntax checking script to improve log handling and streamline error aggregation. Added support for parallel processing and enhanced readability.

wasertech committed:
Added compute_stats function to summarize syntax errors in C++ and header files. Improved count_lines function to return line counts and handle empty files.
